### PR TITLE
Implement correct interactive library tabs

### LINKED ISSUE #252: Implement proper library tabs
Implement proper library tabs by:

- [x] Add correct design and icons
- [x] Add quick-navigation to add a new text element
- [x] Add arrow navigation support when focus is on any tab (any tab will auto-select on focus, so you can't have a focused tab be not selected, which makes ui simpler)
- [x] Add (only currently) selected tab to tab order (technically remove non-selected tabs from tab order)
- [ ] Add global shortcut to quickly focus currently selected tab

### PR DESCRIPTION
**Pull request by barklund**
This implements library tabs as per design and interactive roles as expected.

- [x] Update tab naming
- [x] Add animation tab
- [x] Restructure libraries to use `insertElement` from context rather than from prop
- [x] Implement text quick-action
- [x] Add only active tab to tab order (and if text, also quick-action)
- [x] Add accessible labels to tabs and quick-action
- [x] Add relevant semantic roles where applicable
- [x] Complete refactor to better structure
- [x] Add keyboard arrow navigation to tabs when focused
- [x] Double check focus & hover styles

Fixes #252

### CODE REVIEW

**Comment by spacedmonkey** [comment]
I know this is a small thing, but can you break this into directories for the different tabs. It just makes it easier to review.

File: assets/src/edit-story/components/library/tabs/index.js
Comment (swissspidy) [inline assets/src/edit-story/components/library/tabs/index.js]: Nesting buttons isn't valid HTML and throws `validateDOMNesting` warnings too. That means we need to move it higher up and come up with a way to position the `QuickAction` appropriately.

File: assets/src/edit-story/components/library/tabs/index.js
Comment (swissspidy) [inline assets/src/edit-story/components/library/tabs/index.js]:
```suggestion
    // @todo: Needs product definition
    insertElement('text', {
```

File: assets/src/edit-story/components/library/tabs/index.js
Comment (swissspidy) [inline assets/src/edit-story/components/library/tabs/index.js]: Would need to be translatable

File: assets/src/edit-story/components/library/tabs/index.js
Comment (barklund) [reply assets/src/edit-story/components/library/tabs/index.js]: How about moving the outer functionality to the `<li/>` (the button is semantically irrelevant anyway) and giving it an appropriate interactive role and `onClick`? Would that be proper?

File: assets/src/edit-story/components/library/tabs/index.js
Comment (barklund) [reply assets/src/edit-story/components/library/tabs/index.js]: Ah yes, it's basically just dummy, but might survive all the way to product 😉. Will add i18n.

**Comment by barklund** [comment]
> I know this is a small thing, but can you break this into directories for the different tabs. It just makes it easier to review.

That's actually a good idea, though it would create quite the conflict with #266.

File: assets/src/edit-story/components/library/tabs/index.js
Comment (barklund) [reply assets/src/edit-story/components/library/tabs/index.js]: I have done this now in the refactor. I think it's semantically valid.

**Comment by barklund** [comment]
I'm completely overhauled the internals of the library to a "better" structure, that's easier to extend. The media library pane should probably be split in multiple components now, but that's an easy refactor for later.

PTAL @swissspidy, @spacedmonkey, @dvoytenko.

File: assets/src/edit-story/components/library/libraryTabs.js
Comment (barklund) [inline assets/src/edit-story/components/library/libraryTabs.js]: This was soooo easy to do! Thanks, @dvoytenko! And because of the DOM based model, I don't even need to add anything about "only if focus", because key events will only get to the parent node, if current focus is on one of the tabs.

I don't even see a need to generalize this for other components needing similar but different implementation, as it's so easy.

I didn't find a good way to make it a single callback though. I first examined `evt.key`, but its value is `ArrowLeft` which would be a different syntax than the simple `left` used in setting up the callback. Which wouldI found quite confusing. Is there a way to catch which key has been pressed in MouseTrap lingo? However, using different callbacks works well, so it's not a big issue here.

File: assets/src/edit-story/components/library/panes/elements/elements.svg
Comment (barklund) [inline assets/src/edit-story/components/library/panes/elements/elements.svg:9]: I don't really know why this icon is an image and not vector, but I guess that's for future optimizations.

File: assets/src/edit-story/components/library/libraryTabs.js
Comment (dvoytenko) [reply assets/src/edit-story/components/library/libraryTabs.js]: Unfortunately, only partially possible, I believe. I took a look superficially before. It looks like handler passed to `bind` takes two args: KeyboardEvent and what they call `combo`. The `combo` uses their naming, but it contains modifiers and everything. E.g. it'd be `mod+down`. But I could be wrong.

Comment (dvoytenko) [reply assets/src/edit-story/components/library/libraryTabs.js]: Bottom line: I figured in some other places, I could live with dichotomy for a few special cases such as arrows.

File: assets/src/edit-story/components/library/libraryTabs.js
Comment (dvoytenko) [inline assets/src/edit-story/components/library/libraryTabs.js]: Nit: but I usually find it easier when it's a `dir` number with 1 and -1 values.

File: assets/src/edit-story/components/library/panes/shared.js
Comment (dvoytenko) [inline assets/src/edit-story/components/library/panes/shared.js]: Why both `hidden` attribute and `display:none`?

**Review by dvoytenko — CHANGES_REQUESTED** [review]

**Comment by dvoytenko** [comment]
Overall, I'm at LGTM. Just a couple of nits. Heads up on possible conflicts with #266.

File: assets/src/edit-story/components/library/libraryLayout.js
Comment (swissspidy) [inline assets/src/edit-story/components/library/libraryLayout.js]: I18N ☝️

Comment (barklund) [reply assets/src/edit-story/components/library/libraryLayout.js]: Fixed in f7756942.

File: assets/src/edit-story/components/library/libraryTabs.js
Comment (barklund) [reply assets/src/edit-story/components/library/libraryTabs.js]: Fixed in f7756942.

File: assets/src/edit-story/components/library/panes/shared.js
Comment (barklund) [reply assets/src/edit-story/components/library/panes/shared.js]: DIdn't actually know that browsers hid any element that was marked with `hidden`. But it does, so I removed it in f7756942. Thanks!

File: assets/src/edit-story/components/library/libraryTabs.js
Comment (barklund) [reply assets/src/edit-story/components/library/libraryTabs.js]: I'll leave it as is for now.

File: assets/src/edit-story/components/library/panes/shared.js
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/shared.js:35]: This should really be a button to denote to screen readers it is clickable.

File: assets/src/edit-story/components/library/panes/shared.js
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/shared.js:37]: We could also add `aria-selected` here. 
```
aria-selected: isActive,
```

File: assets/src/edit-story/components/library/panes/shared.js
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/shared.js]: Screenshot 2020-02-24 at 10 38 19
This focus ring looks really change. It should really be 100% width and height.

**Comment by spacedmonkey** [comment]
There are existing tabs with some good accessibility can be found [here](https://github.com/google/web-stories-wp/blob/40cdc3e9f92b4569c91b229fec8bfe6e9f09c8e3/assets/src/edit-story/components/inspector/inspectorTabs.js#L85-L92).

File: assets/src/edit-story/components/library/panes/shared.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/shared.js:35]: See example [here](https://github.com/google/web-stories-wp/blob/40cdc3e9f92b4569c91b229fec8bfe6e9f09c8e3/assets/src/edit-story/components/inspector/inspectorTabs.js#L85-L92) and follows this [example](https://www.w3.org/TR/wai-aria-practices/examples/tabs/tabs-1/tabs.html).

File: assets/src/edit-story/components/library/panes/elements/elementsPane.js
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/panes/elements/elementsPane.js]: Why hard code the id here? Can't we have a helper function to generate this id as they all follow the same pattern.

File: assets/src/edit-story/components/library/libraryProvider.js
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/libraryProvider.js:31]: FYI for drag and drop into page, it needs to insert element into the page. I want to move this the `insertElement` method to somewhere easier to get at global. Can we do this as part of this refactor as well.

**Review by spacedmonkey — CHANGES_REQUESTED** [review]
Some accessibility issues and code tidying required.

File: assets/src/edit-story/components/library/panes/shared.js
Comment (barklund) [reply assets/src/edit-story/components/library/panes/shared.js:37]: Ah yes, that should be added.

File: assets/src/edit-story/components/library/panes/shared.js
Comment (barklund) [reply assets/src/edit-story/components/library/panes/shared.js:35]: It has [the interactive role `tab` which denotes to screen readers, that it's clickable](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/Tab_Role). As discussed [here](https://github.com/google/web-stories-wp/pull/361#discussion_r382576516), I think this is the best semantic solution for this widget.

File: assets/src/edit-story/components/library/panes/shared.js
Comment (barklund) [reply assets/src/edit-story/components/library/panes/shared.js]: Then it'll be cut off by surrounding elements and you'll only see a small part of it. Maybe only 1, 2 or 3 sides, which makes it visually unusable.

I am very open to a better focus state in UX, but barring that, I implemented what I thought was most functional.

File: assets/src/edit-story/components/library/panes/elements/elementsPane.js
Comment (barklund) [reply assets/src/edit-story/components/library/panes/elements/elementsPane.js]: Fixed in 4be495bd. And yes, I knew I should probably generalize when I made it ;)

File: assets/src/edit-story/components/library/libraryProvider.js
Comment (barklund) [reply assets/src/edit-story/components/library/libraryProvider.js:31]: Hm, it doesn't actually do a lot more than `addElement` from the global story provider besides capping the width and height and converting media stuff to background elements. But it does do some logic, so it's probably a valid request. I wouldn't really know where to place it globally even.

Alternatively, drag-n-drop could just use the same `useInsertElement` hook and we could move **that** to a more general location?

File: assets/src/edit-story/components/library/panes/shared.js
Comment (barklund) [reply assets/src/edit-story/components/library/panes/shared.js:37]: Fixed in 4be495bd.

File: assets/src/edit-story/components/library/libraryProvider.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/libraryProvider.js:31]: For reference. https://github.com/google/web-stories-wp/pull/376

**Review by dvoytenko — APPROVED** [review]
LGTM from me.

File: assets/src/edit-story/components/library/panes/shared.js
Comment (spacedmonkey) [reply assets/src/edit-story/components/library/panes/shared.js]: @samitron7 ^ Flagged for UX review.

File: assets/src/edit-story/components/library/useInsertElement.js
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/useInsertElement.js:24]: In my PR [here](https://github.com/google/web-stories-wp/pull/376), I moved this file to `edit-story/app/media/actions/useInsertMediaElement.js`

**Review by swissspidy — APPROVED** [review]
LGTM too. Just added a small todo note to one part, mostly for myself.

**Comment by swissspidy** [comment]
@spacedmonkey Have your last comments been appropriately addressed? Any objections to merging this as-is?

**Comment by barklund** [comment]
Let's await focus and hover styles from UX before merging. @samitron7 has promised it within a day.

File: assets/src/edit-story/components/library/libraryTabs.js
Comment (spacedmonkey) [inline assets/src/edit-story/components/library/libraryTabs.js:52]: For reference #351

**Review by spacedmonkey — DISMISSED** [review]

File: assets/src/edit-story/components/library/libraryTabs.js
Comment (barklund) [reply assets/src/edit-story/components/library/libraryTabs.js:52]: Which change are you specifically requesting here?

File: assets/src/edit-story/components/library/panes/shared.js
Comment (samitron7) [reply assets/src/edit-story/components/library/panes/shared.js]: @spacedmonkey @barklund I'm currently working on all the hover states now. Stay tune. Will flag when it's done.

File: assets/src/edit-story/components/library/libraryTabs.js
Comment (swissspidy) [reply assets/src/edit-story/components/library/libraryTabs.js:52]: No change requested :-) Just an FYI that there‘s a related PR

Comment (barklund) [reply assets/src/edit-story/components/library/libraryTabs.js:52]: It's just that the github issue says _"@spacedmonkey requested changes"_ with only this comment... And this change request is blocking merging (besides the outstanding UX question), so I just wanted to check if that was a mistake or not.

Screen Shot 2020-02-26 at 3 13 56 PM

Comment (swissspidy) [reply assets/src/edit-story/components/library/libraryTabs.js:52]: Dismissed the review now 🙂

File: assets/src/edit-story/components/library/panes/shared.js
Comment (barklund) [reply assets/src/edit-story/components/library/panes/shared.js]: For now the entire tab changes color to `colors.action` to stand out as something clearly missing. Created followup in #423.

Current temporary design for selected tab (with and without focus):
Screen Shot 2020-02-27 at 2 29 16 PM
Screen Shot 2020-02-27 at 2 29 27 PM

File: assets/src/edit-story/components/library/panes/media/mediaTab.js
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/media/mediaTab.js]: Wrong label?

File: assets/src/edit-story/components/library/panes/shapes/shapesTab.js
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/shapes/shapesTab.js]: Wrong label?

File: assets/src/edit-story/components/library/panes/text/textTab.js
Comment (swissspidy) [inline assets/src/edit-story/components/library/panes/text/textTab.js]: Are both `stopPropagation ` and `preventDefault` needed here?

Comment (barklund) [reply assets/src/edit-story/components/library/panes/text/textTab.js]: Good point. `preventDefault` isn't strictly necessary here, removed in a24838a0 along with measures to remove some console errors arising from unwanted props leaking through to components. Thanks for flagging!

File: assets/src/edit-story/components/library/panes/media/mediaTab.js
Comment (barklund) [reply assets/src/edit-story/components/library/panes/media/mediaTab.js]: Fixed in 8f8268b5.

File: assets/src/edit-story/components/library/panes/shapes/shapesTab.js
Comment (barklund) [reply assets/src/edit-story/components/library/panes/shapes/shapesTab.js]: Fixed in 8f8268b5.

**Pull request merged**
